### ISSUE TITLE
elastic-apm-node with async hooks enabled leaks Transactions when used with cls-rtracer

### ISSUE THREAD

**Issue by tsondergaard** [elastic/apm-agent-nodejs#1807]
elastic-apm-node leaks Transaction objects when it is used in an express app that uses `require("cls_rtracer").expressMiddleware()`. 

**To Reproduce**

Steps to reproduce the behavior: 

1. Unzip [cls-rtracer-example.zip](https://github.com/elastic/apm-agent-nodejs/files/5042256/cls-rtracer-example.zip)
2. Cd to cls-rtracer-example directory
3. Run `npm ci`
4. Run `ELASTIC_APM_SERVICE_NAME=cls-rtracer-example ELASTIC_APM_SERVER_URL=http://elastic-server:8200 node --require elastic-apm-node/start index.js`
4. Hit http://localhost:3000 a bunch of time in a browser or with curl

**Expected behavior**

Following these steps the `activeTransactions` map in node_modules/elastic-apm-node/lib/instrumentation/async-hooks.js will continue to grow when the server processes http requests until the node process eventually runs out of memory.

**Environment (please complete the following information)**

- OS: Fedora Linux 31 x86_64
- Node.js version: Reproduced with 12.18.3 and 14.7.0
- APM Server version: 8.5
- Agent version: 3.6.1

**How are you starting the agent? (please tick one of the boxes)**

- [ ] Calling `agent.start()` directly (e.g. `require('elastic-apm-node').start(...)`)
- [ ] Requiring `elastic-apm-node/start` from within the source code
- [x] Starting node with `-r elastic-apm-node/start`

**Additional context**

I believe it is important to use cls-rtracer 2.x to reproduce this bug as it uses the [AsyncLocalStorage API](https://nodejs.org/api/async_hooks.html#async_hooks_class_asynclocalstorage) that is part of the async_hooks module in recent versions of node.

- Agent config options 
  <details>
    <summary>Click to expand</summary>

    ```
    ELASTIC_APM_SERVICE_NAME=cls-rtracer-example
    ELASTIC_APM_SERVER_URL=http://elastic-server:8200
    ```
  </details>
- `package.json` dependencies:
  <details>
    <summary>Click to expand</summary>

    ```
    "dependencies": {
        "cls-rtracer": "^2.2.0",
        "elastic-apm-node": "^3.6.1",
        "express": "^4.17.1"
    }
    ```
  </details>

**Comment by tsondergaard** [comment]
By setting breakpoints in the program in cls-rtracer-example.zip in node_modules/cls-rtracer/src/rtracer.js:12 and node_modules/elastic-apm-node/lib/instrumentation/async-hooks.js:28, I can see that the executionAsyncId() of the leaked Transactions are from the new AsyncResource('cls-rtracer') instantiated in node_modules/cls-rtracer/src/rtracer.js:12. I'm not sure why that AsyncResource needs to be created in cls-rtracer, perhaps @puzpuzpuz can help?

**Comment by puzpuzpuz** [comment]
@tsondergaard `cls-rtracer` creates an `AsyncResource` when monkey-patching EventEmitters (`req` and `res`) in order to guarantee proper async context for event listeners that another middleware or user code might install. You can see the corresponding code here: https://github.com/puzpuzpuz/cls-rtracer/blob/master/src/util.js#L8

I'm not sure how this can create a leak, unless elastic-apm-node's CLS implementation does some kind of additional tracking for custom `AsyncResource`. But I didn't check elastic-apm-node's code, so I'm not sure what exactly goes wrong here.

In general, it's weird for me to see that someone may combine `cls-rtracer` and an APM agent. My understanding is that APM agents should provide request (including logging) capabilities for the end application, as they have all the infrastructure in place. Am I wrong?

**Comment by tsondergaard** [comment]
Thanks for replying, @puzpuzpuz. The application I am working on was already using cls-rtracer to include X-Correlation-Id in a log file written with winston. Adding elastic-apm-node is new and I did not consider that they would interact poorly. The systems that are calling my service do not necessarily use elastic APM, so I would like to continue to use cls-rtracer to log X-Correlation-Id in my winston logs.

**Comment by puzpuzpuz** [comment]
@tsondergaard thanks for the clarification. Now it makes sense.

As `cls-rtracer` doesn't do anything unusual or potentially harmful, let's wait for some inputs from elastic APM agent maintainers.

**Comment by tsondergaard** [comment]
I have found out the reason for the leak. cls-rtracer binds an AsyncResource to http req and res objects:

```
const wrapHttpEmitters = (req, res) => {
  const asyncResource = new AsyncResource('cls-rtracer')
  wrapEmitter(req, asyncResource)
  wrapEmitter(res, asyncResource)
}
```

elastic-apm-node attaches req and res to a Transaction object that it stored in a map and it deletes that transaction object when the async_hook destroy callback is invoked. It never will be as the AsyncResource remains alive.

@puzpuzpuz , could cls-rtracer destroy the AsyncResource earlier, e.g when the response is finished?

![elastic-apm-node-leak](https://user-images.githubusercontent.com/2016842/90401009-a622b080-e09d-11ea-8dfd-0dba2ded14ef.png)

**Comment by puzpuzpuz** [comment]
@tsondergaard AsyncResource should trigger destroy hook once it gets GCed, which should happen once req/res objects are GCed. So, there is no need to trigger the event "manually". If req/res get promoted to the old gen, it may take a while to happen, but GC should kick in sooner or later.

Some more details on Node.js core implementation of async_hooks: if elastic agent registers its async hook before cls-rtracer processes any HTTP request (which should be the case), all AsyncResources should be subject for GC tracking and trigger the destroy hook.

**Comment by tsondergaard** [comment]
@puzpuzpuz, the point of my previous post is that elastic-apm-node maintains a map of active transactions. The Transaction objects stored in this map have res and req properties and with cls-rtracer, the res and req objects have emit methods that capture the AsyncResource. The Transaction object will be removed from the map when the AsyncResource is destroyed (via the async_hooks destroy callback). But this will never happen as the AsyncResource object is referenced.

In other words, elastic-apm-node keeps the AsyncResource object alive via its activeTransactions map and will not remove its reference to it until the AsyncResource is garbage collected.

**Comment by puzpuzpuz** [comment]
@tsondergaard I see now. Going to take a look at the reproducer tomorrow.

**Comment by puzpuzpuz** [comment]
@tsondergaard I've checked the reproducer and it seems that current `wrapHttpEmitters` implementation in `cls-rtracer` (which monkey-patches `EE.emit()` method) conflicts with elastic-apm-node CLS related logic. As the result not all `Transaction` objects are removed from the `activeTransactions` map. I've changed the logic, so that the problem should be gone once this PR passes review: https://github.com/puzpuzpuz/cls-rtracer/pull/39

Thanks for reporting this issue.
